File: ios/App/Snappet/Features/Expense/ReceiptParser.swift
import Foundation

/// Pure, device-free parser that turns pasted/scanned receipt text into line items plus
/// the detected tax, discount and total. Heuristic by necessity — receipts vary wildly —
/// but the rules are simple and unit-tested against a real Costco receipt:
///
///  - A line item is a row ending in a money amount (e.g. `LIQUIDIV LLS  28.99 E`). The
///    last money token on the line is the price; leading item-codes (bare integers) and a
///    leading single-letter tax flag are stripped from the name.
///  - A trailing-minus amount (`4.00-`) is an instant-savings/discount line and is summed
///    into `discount` rather than added as an item.
///  - Summary rows — SUBTOTAL / TAX / TOTAL / payment / membership lines — are not items;
///    the tax and grand-total values are captured from them when present.
///
/// Kept free of SwiftUI/SwiftData so it runs in `SnappetTests` with no simulator.
enum ReceiptParser {

    struct ParsedReceipt: Equatable {
        var items: [ReceiptItem]
        var discount: Double
        var tax: Double?
        var total: Double?
    }

    /// Keyword fragments that mark a line as a receipt summary / metadata row rather than a
    /// purchasable item. Matched case-insensitively as substrings.
    private static let skipKeywords = [
        "SUBTOTAL", "TOTAL", "TAX", "CHANGE", "BALANCE", "VISA", "MASTERCARD",
        "DEBIT", "CREDIT", "AMOUNT", "INSTANT SAVINGS", "MEMBER", "CHIP READ",
        "APPROVED", "RESP", "TRAN", "AID", "SEQ", "APP#", "ITEMS SOLD", "FSA",
        "COUNT", "BOTTOM OF BASKET", "THANK YOU", "PLEASE COME", "OP#", "WHSE",
        "TRM", "TRN", "WHOLESALE", "PURCHASE", "ITEM",
    ]

    static func parse(_ text: String) -> ParsedReceipt {
        var items: [ReceiptItem] = []
        var discount = 0.0
        var tax: Double?
        var total: Double?

        for rawLine in text.split(whereSeparator: \.isNewline) {
            let line = String(rawLine).trimmingCharacters(in: .whitespaces)
            guard !line.isEmpty else { continue }
            let upper = line.uppercased()

            guard let money = lastMoney(in: line) else { continue }

            // Discount / instant-savings rows (trailing minus) — credit, don't itemize.
            if money.isNegative {
                discount += money.value
                continue
            }

            // Summary / metadata rows: capture tax & grand total, never itemize.
            if skipKeywords.contains(where: { upper.contains($0) }) {
                // "TOTAL TAX 14.01" or a line that is just the tax.
                if upper.contains("TAX") && !upper.contains("N/TAX") {
                    tax = money.value
                } else if upper.contains("TOTAL")
                            && !upper.contains("SUBTOTAL")
                            && !upper.contains("NUMBER")
                            && !upper.contains("BOB") {
                    // Prefer the largest total seen (the grand total, not a card line).
                    total = max(total ?? 0, money.value)
                }
                continue
            }

            let name = cleanName(line, priceToken: money.token)
            guard !name.isEmpty else { continue }
            items.append(ReceiptItem(name: name, price: money.value))
        }

        return ParsedReceipt(items: items, discount: discount, tax: tax, total: total)
    }

    // MARK: - Money scanning

    private struct Money { let value: Double; let isNegative: Bool; let token: String }

    /// The last money-looking token on a line (e.g. `28.99`, `$1,234.50`, `4.00-`).
    private static func lastMoney(in line: String) -> Money? {
        var found: Money?
        for token in line.split(whereSeparator: { $0 == " " || $0 == "\t" }) {
            if let money = money(from: String(token)) { found = money }
        }
        return found
    }

    private static func money(from token: String) -> Money? {
        var t = token
        // Drop a trailing single tax-flag letter (Costco appends A/B/E/F to amounts,
        // sometimes attached: "28.99E", "4.00-A").
        if let last = t.last, last.isLetter, t.count > 1 { t.removeLast() }
        let isNegative = t.hasSuffix("-")
        if isNegative { t.removeLast() }
        t = t.replacingOccurrences(of: "$", with: "")
             .replacingOccurrences(of: ",", with: "")
        // Require an explicit two-decimal money shape so bare item-codes aren't prices.
        guard let dot = t.firstIndex(of: "."),
              t.distance(from: dot, to: t.endIndex) == 3,
              t.dropFirst(t.distance(from: t.startIndex, to: dot) + 1).allSatisfy(\.isNumber),
              t[..<dot].allSatisfy(\.isNumber),
              !t[..<dot].isEmpty,
              let value = Double(t) else { return nil }
        return Money(value: value, isNegative: isNegative, token: token)
    }

    /// Strip everything from the price token onward, then drop a leading single-letter tax
    /// flag and any leading bare item-code integers, leaving the product description.
    private static func cleanName(_ line: String, priceToken: String) -> String {
        var head = line
        if let range = line.range(of: priceToken) {
            head = String(line[..<range.lowerBound])
        }
        var tokens = head.split(whereSeparator: { $0 == " " || $0 == "\t" }).map(String.init)
        // Leading single-letter tax flag (Costco prefixes some rows with E/F).
        if let first = tokens.first, first.count == 1, first.allSatisfy(\.isLetter) {
            tokens.removeFirst()
        }
        // Leading bare item-codes (pure digits, possibly with a stray '/').
        while let first = tokens.first,
              first.allSatisfy({ $0.isNumber || $0 == "/" }),
              first.contains(where: \.isNumber) {
            tokens.removeFirst()
        }
        return tokens.joined(separator: " ").trimmingCharacters(in: .whitespaces)
    }
}

File: android/app/src/main/java/com/snappet/mobile/feature/expense/ReceiptParser.kt
package com.snappet.mobile.feature.expense

/**
 * Pure, device-free parser that turns pasted / OCR'd receipt text into line items plus the
 * detected tax, discount and total. Ported 1:1 from the iOS `ReceiptParser`. Heuristic but
 * unit-tested against real Costco lines: a row ending in a money amount is an item (leading
 * item-codes and a trailing tax-flag letter are stripped); a trailing-minus amount is an
 * instant-savings/discount; SUBTOTAL / TAX / TOTAL / payment rows are not items.
 */
object ReceiptParser {

    data class ParsedReceipt(
        val items: List<ReceiptItem>,
        val discount: Double,
        val tax: Double?,
        val total: Double?,
    )

    private val skipKeywords = listOf(
        "SUBTOTAL", "TOTAL", "TAX", "CHANGE", "BALANCE", "VISA", "MASTERCARD",
        "DEBIT", "CREDIT", "AMOUNT", "INSTANT SAVINGS", "MEMBER", "CHIP READ",
        "APPROVED", "RESP", "TRAN", "AID", "SEQ", "APP#", "ITEMS SOLD", "FSA",
        "COUNT", "BOTTOM OF BASKET", "THANK YOU", "PLEASE COME", "OP#", "WHSE",
        "TRM", "TRN", "WHOLESALE", "PURCHASE", "ITEM",
    )

    fun parse(text: String): ParsedReceipt {
        val items = mutableListOf<ReceiptItem>()
        var discount = 0.0
        var tax: Double? = null
        var total: Double? = null

        for (rawLine in text.split('\n', '\r')) {
            val line = rawLine.trim()
            if (line.isEmpty()) continue
            val upper = line.uppercase()

            val money = lastMoney(line) ?: continue

            // Discount / instant-savings rows (trailing minus) — credit, don't itemize.
            if (money.isNegative) {
                discount += money.value
                continue
            }

            // Summary / metadata rows: capture tax & grand total, never itemize.
            if (skipKeywords.any { upper.contains(it) }) {
                if (upper.contains("TAX") && !upper.contains("N/TAX")) {
                    tax = money.value
                } else if (upper.contains("TOTAL") &&
                    !upper.contains("SUBTOTAL") &&
                    !upper.contains("NUMBER") &&
                    !upper.contains("BOB")
                ) {
                    total = maxOf(total ?: 0.0, money.value)
                }
                continue
            }

            val name = cleanName(line, money.token)
            if (name.isEmpty()) continue
            items.add(ReceiptItem(name, money.value))
        }

        return ParsedReceipt(items, discount, tax, total)
    }

    private data class Money(val value: Double, val isNegative: Boolean, val token: String)

    private fun lastMoney(line: String): Money? {
        var found: Money? = null
        for (token in line.split(' ', '\t').filter { it.isNotEmpty() }) {
            money(token)?.let { found = it }
        }
        return found
    }

    private fun money(token: String): Money? {
        var t = token
        // Drop a trailing single tax-flag letter (Costco: "28.99E", "4.00-A").
        if (t.length > 1 && t.last().isLetter()) t = t.dropLast(1)
        val isNegative = t.endsWith("-")
        if (isNegative) t = t.dropLast(1)
        t = t.replace("$", "").replace(",", "")

        val dot = t.indexOf('.')
        if (dot < 0) return null
        if (t.length - dot != 3) return null // require exactly two decimals
        val intPart = t.substring(0, dot)
        val decPart = t.substring(dot + 1)
        if (intPart.isEmpty() || !intPart.all { it.isDigit() } || !decPart.all { it.isDigit() }) return null
        val value = t.toDoubleOrNull() ?: return null
        return Money(value, isNegative, token)
    }

    private fun cleanName(line: String, priceToken: String): String {
        val idx = line.indexOf(priceToken)
        val head = if (idx >= 0) line.substring(0, idx) else line
        val tokens = head.split(' ', '\t').filter { it.isNotEmpty() }.toMutableList()
        // Leading single-letter tax flag (Costco prefixes some rows with E/F).
        if (tokens.isNotEmpty() && tokens[0].length == 1 && tokens[0].all { it.isLetter() }) {
            tokens.removeAt(0)
        }
        // Leading bare item-codes (pure digits, possibly with a stray '/').
        while (tokens.isNotEmpty() && tokens[0].all { it.isDigit() || it == '/' } && tokens[0].any { it.isDigit() }) {
            tokens.removeAt(0)
        }
        return tokens.joinToString(" ").trim()
    }
}
